go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+package vtgate
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+	pbq "github.com/youtube/vitess/go/vt/proto/query"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/tabletserver/tabletconn"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// AddrUpdate updates about an endpoint and whether it is up.
+type AddrUpdate struct {
+	EndPoint *pbt.EndPoint
+	IsUp     bool
+}
+
+// NewHealthCheck creates a new HealthCheck object.
+func NewHealthCheck(endPoints []*pbt.EndPoint, addrUpdate <-chan *AddrUpdate, connTimeout time.Duration, retryDelay time.Duration) *HealthCheck {
+	hc := &HealthCheck{
+		addrToConns: make(map[string]*healthCheckConn),
+		targetToEPs: make(map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint),
+		connTimeout: connTimeout,
+		retryDelay:  retryDelay,
+	}
+	hc.initConns(endPoints)
+	go hc.onAddrUpdate(addrUpdate)
+	return hc
+}
+
+// HealthCheck performs health checking and notifies downstream components about any changes.
+type HealthCheck struct {
+	mu          sync.RWMutex
```

Comment:
Please add a comment to say what the mutex protects (the map) and doesn't protect (the parameters)